modules/cdp-engine/src/lib.rs

```diff
@@ -1215,59 +1269,207 @@ impl<T: Config> Pallet<T> {
-			}
+		let (actual_supply_collateral, actual_target_amount) = <T as Config>::CDPTreasury::swap_collateral_to_stable(
+			currency_id,
+			SwapLimit::ExactTarget(collateral_supply, target_stable_amount),
+			false,
+		)?;
 
-			// Note: for StableAsset, the swap of cdp treasury is always on `ExactSupply`
-			// regardless of this swap_limit params. There will be excess stablecoins that
-			// need to be returned to the `who` from cdp treasury account.
-			if actual_target_amount > target_stable_amount {
-				<T as Config>::CDPTreasury::withdraw_surplus(
-					who,
-					actual_target_amount.saturating_sub(target_stable_amount),
-				)?;
-			}
-		} else {
-			// if cannot liquidate by swap, create collateral auctions by cdp treasury
-			<T as Config>::CDPTreasury::create_collateral_auctions(
-				currency_id,
-				amount,
-				target_stable_amount,
-				who.clone(),
-				true,
+		let refund_collateral_amount = amount
+			.checked_sub(actual_supply_collateral)
+			.expect("swap success means collateral >= actual_supply_collateral; qed");
+		// refund remain collateral to CDP owner
+		if !refund_collateral_amount.is_zero() {
+			<T as Config>::CDPTreasury::withdraw_collateral(who, currency_id, refund_collateral_amount)?;
+		}
+
+		// Note: for StableAsset, the swap of cdp treasury is always on `ExactSupply`
+		// regardless of this swap_limit params. There will be excess stablecoins that
+		// need to be returned to the `who` from cdp treasury account.
+		if actual_target_amount > target_stable_amount {
+			<T as Config>::CDPTreasury::withdraw_surplus(
+				who,
+				actual_target_amount.saturating_sub(target_stable_amount),
 			)?;
 		}
 
 		Ok(())
 	}
-	pub fn get_collateral_currency_ids() -> Vec<CurrencyId> {
-		CollateralParams::<T>::iter_keys().collect()
+}
+
+pub struct LiquidateViaContracts<T>(PhantomData<T>);
+impl<T: Config> LiquidateCollateral<T::AccountId> for LiquidateViaContracts<T> {
+	fn liquidate(
+		who: &T::AccountId,
+		currency_id: CurrencyId,
+		amount: Balance,
+		target_stable_amount: Balance,
+	) -> DispatchResult {
+		let liquidation_contracts = Pallet::<T>::liquidation_contracts();
+		let liquidation_contracts_len = liquidation_contracts.len();
+		if liquidation_contracts_len.is_zero() {
+			return Err(Error::<T>::LiquidationFailed.into());
+		}
+
+		let max_supply_limit = Ratio::one()
+			.saturating_sub(T::MaxLiquidationContractSlippage::get())
+			.reciprocal()
+			.unwrap_or_else(Ratio::max_value)
+			.saturating_mul_int(
+				T::PriceSource::get_relative_price(T::GetStableCurrencyId::get(), currency_id)
+					.expect("the oracle price should be available because liquidation are triggered by it.")
+					.saturating_mul_int(target_stable_amount),
+			);
+		let collateral_supply = amount.min(max_supply_limit);
+
+		let collateral = currency_id
+			.erc20_address()
+			.ok_or(Error::<T>::CollateralContractNotFound)?;
+		let repay_dest = Pallet::<T>::evm_address();
+		let repay_dest_account_id = Pallet::<T>::account_id();
```

Comment:
Here we are using CdpEngine's pallet account as the repayment account for contract liquidation. Is there a specific reason for this instead of using the Treasury's account? 
Since the Treasury has the collateral and the debt, would it be easier if we use Treasury's account here instead?

Reply:
I don't see how it gets "easier" using treasury account. Mixing up all balances in one treasury account will make it difficult to tell what goes wrong if anything unexpected happened.